module/observations/helper.js

```diff
@@ -2819,20 +2823,53 @@ module.exports = class ObservationsHelper {
           let topLevelEntityId = bodyData[topLevelEntityType];
 
 
-          let roles = bodyData.role.split(",");
+          let tenantDetails = await userService.fetchPublicTenantDetails(tenantData.tenantId);
+          if (!tenantDetails.data || !tenantDetails.data.meta || tenantDetails.success !== true) {
+            throw {
+              status: httpStatusCode.bad_request.status,
+              message: messageConstants.apiResponses.FAILED_TO_FETCH_TENANT_DETAILS,
+            };
+          }
+
+          let observableEntityKeys = tenantDetails.data.meta.observableEntityKeys;
+
+          if(observableEntityKeys === undefined || observableEntityKeys === null || observableEntityKeys.length === 0){
+            throw {
+              status: httpStatusCode.bad_request.status,
+              message: messageConstants.apiResponses.OBSERVABLE_ENTITY_KEYS_NOT_FOUND,
+            };
+          }
+          
+          let KyetoValidate = [];
+
+          for(let i=0;i<observableEntityKeys.length;i++){
+              KyetoValidate.push(observableEntityKeys[i]);
+          }
+
+          let roles = [];
+          for(let i=0;i<KyetoValidate.length;i++){
+            if(bodyData[KyetoValidate[i]]){
+              let rolesFound = bodyData[KyetoValidate[i]]
+              let roleElements = rolesFound.split(',')
+
+              for(let j=0;j<roleElements.length;j++){
+                roles.push(roleElements[j]);
+              }
+            }
+          }
 
           let entityTypeArr = [];
+
           for(let roleIndex=0;roleIndex<roles.length;roleIndex++){
 
-          let rolesDocumentAPICall = await entityManagementService.userRoleExtension({
-            code: roles[roleIndex],
+          let rolesDocumentAPICall = await entityManagementService.entityDocuments({
+            _id: roles[roleIndex],
             tenantId:solutionDocument[0].tenantId,
-            orgIds:{$in:['ALL',...solutionDocument[0].orgIds]}
-          },
-          ["entityTypes.entityType"])
+            orgIds:{$in:['ALL',solutionDocument[0].orgId]}
+          },["metaInformation.targetedEntityTypes"]);
 
           if(rolesDocumentAPICall.success){
```

Comment:
@borkarsaish65  validate and optimize
if (
      rolesDocumentAPICall?.success &&
      Array.isArray(rolesDocumentAPICall.data) &&
      rolesDocumentAPICall.data[0]?.metaInformation?.targetedEntityTypes?.[0]?.entityType
    ) {
      entityTypeArr.push(
        rolesDocumentAPICall.data[0].metaInformation.targetedEntityTypes[0].entityType
      );
    } else {
      throw {
        status: httpStatusCode.bad_request.status,
        message: messageConstants.apiResponses.USER_ROLES_NOT_FOUND,
      };
    }